pkg/network/protocols/http/ebpf_gotls.go

```diff
@@ -289,92 +290,119 @@ func (p *GoTLSProgram) handleProcessStart(pid pid) {
 		return
 	}
 
-	hookedBin := p.getHookedBinary(stat.Ino)
-	if hookedBin == nil {
-		hookedBin, err = p.hookNewBinary(binPath, stat.Ino)
-		if err != nil {
-			log.Debugf("could not hook new binary (%d): %s", pid, err)
-			return
-		}
-		hookedBin.mTime = stat.Mtim
-		p.setHookedBinary(stat.Ino, hookedBin)
-	} else if stat.Mtim != hookedBin.mTime {
-		log.Warnf("binary %q has been modified since it has been hooked, skipping process registration.", binPath)
-		return
-	}
-
-	p.lock.Lock()
-	defer p.lock.Unlock()
-	hookedBin.runningProcesses[pid] = struct{}{}
-	p.pidToIno[pid] = stat.Ino
-}
-
-func (p *GoTLSProgram) handleProcessStop(pid pid) {
-	p.lock.Lock()
-	defer p.lock.Unlock()
-
-	ino, ok := p.pidToIno[pid]
-	if !ok {
-		return
-	}
-	delete(p.pidToIno, pid)
-
-	hookedBin, present := p.hookedBinaries[ino]
-	if !present {
-		log.Error("could not retrieve hooked binary")
+	oldProcCount, bin, err := p.registerProcess(stat.Ino, pid, stat.Mtim)
+	if err != nil {
+		log.Warnf("could not register new process (%d) with binary %q: %s", pid, binPath, err)
 		return
 	}
 
-	delete(hookedBin.runningProcesses, pid)
-	if len(hookedBin.runningProcesses) == 0 {
-		log.Debugf("no processes left for ino %d", ino)
-		p.unhookBinary(ino)
+	if oldProcCount == 0 {
+		// This is a slow process so let's not halt the watcher while we
+		// are doing this.
+		go p.hookNewBinary(binPath, stat.Ino, pid, bin)
 	}
 }
 
-func (p *GoTLSProgram) hookNewBinary(binPath string, ino inodeNumber) (*hookedBinary, error) {
+func (p *GoTLSProgram) hookNewBinary(binPath string, ino inodeNumber, pid pid, bin *runningBinary) {
```

Comment:
create a new object as a key, as we discuss offline like binaryInfoKey {
dev_id uint64
inode uint64
}